distributed/tests/test_core.py

```diff
@@ -831,3 +831,50 @@ async def test_connection_pool_detects_remote_close():
     # while creating conn2:
     p._validate()
     await p.close()
+
+
+@pytest.mark.asyncio
+async def test_close_properly():
+    """
+    If the server is closed we should cancel all still ongoing coros and close
+    all listeners.
+    GH4704
+    """
+
+    async def sleep(comm=None):
+        # We want to ensure this is actually canceled therefore don't give it a
+        # chance to actually complete
+        await asyncio.sleep(2000000)
+
+    server = await Server({"sleep": sleep})
+    assert server.status == Status.running
+    ports = [8881, 8882, 8883]
+
+    # Previously we close *one* listener, therefore ensure we always use more
+    # than one for this test
+    assert len(ports) > 1
+    for port in ports:
+        await server.listen(port)
+    # We use TCP here for simplicity. If they are closed we should expect other
+    # backends to close properly as well
+    ip = get_ip()
+    rpc_addr = f"tcp://{ip}:{ports[-1]}"
+    async with rpc(rpc_addr) as remote:
+
+        comm = await remote.live_comm()
+        await comm.write({"op": "sleep"})
+        await asyncio.sleep(0.05)
```

Comment:
Instead of sleeping for 0.05 seconds for the `op` to start up, I wonder if there's something more direct we can probe to make this test a bit more robust. For example, waiting until `server._ongoing_coroutines` is populated

```diff
diff --git a/distributed/tests/test_core.py b/distributed/tests/test_core.py
index f5ef0918..6901a907 100644
--- a/distributed/tests/test_core.py
+++ b/distributed/tests/test_core.py
@@ -862,11 +862,11 @@ async def test_close_properly():
     async with rpc(rpc_addr) as remote:

         comm = await remote.live_comm()
+        assert not server._ongoing_coroutines
         await comm.write({"op": "sleep"})
-        await asyncio.sleep(0.05)
-        assert len(server._ongoing_coroutines)
-        listeners = server.listeners
-        assert len(listeners) == len(ports)
+        while not server._ongoing_coroutines:
+            await asyncio.sleep(0.01)
+        assert len(server.listeners) == len(ports)

         for port in ports:
             await assert_can_connect(f"tcp://{ip}:{port}")
```